### ISSUE TITLE
[FEAT] implement CDAE

### ISSUE THREAD

**Issue by twndus** [twndus/YelpRecommendation#5]
## Background
- [CDAE](https://alicezheng.org/papers/wsdm16-cdae.pdf)

## To do
- [x] implements CDAEDataPipeline.py
- [x] implements CDAEDataset.py
- [x] implements negative sampling function for each training iterations.
- [x] implements CDAEModel.py
- [x] implements CDAETrainer.py
- [x] Adjust existing code such as BaseModel, train.py

**Comment by twndus** [comment]
train val test split

train: .72, valid: .08, test: .2

training
  input: train
  loss: train+ns

validation
  input: train
  loss: valid+ns(train+valid)
  evalution: train-masking, valid-topn

evaluation
  input: train+valid
  output: train+valid-masking, test-topn


1. arrange positive sample
2. split into train, valid, test sets
3. mask: [0101], candidates: [1, 31]
  train: train mask + negative candidates (idx)
  valid: train mask + valid mask + negative candidates (idx)
  test : train+valid mask, test mask

**Comment by twndus** [comment]
Referring to the CDAE publication, it seems necessary to implement a negative sampling function for our training iterations.

**Comment by twndus** [comment]
The loss is decreasing, but the ranking metric consistently shows zero. This could be due to an issue with the data or an error in the metric calculation logic.

```
2024-05-08 05:42:34.893 | INFO     | trainers.base_trainer:run:73 - [Trainer] epoch: 9 > train loss: 1.9184522463474423 / 
                        valid loss: 2.501233343500644 / 
                        precision@K : 0.0 / 
                        Recall@K: 0.0 / 
                        MAP@K: 0.0 / 
                        NDCG@K: 0.0
```

**Comment by GangBean** [comment]
The problem @twndus mentioned caused by below code.
The actual set of validate should be based on valid mask, but we used wrong mask, input_mask(train_mask).
```
# argpartition top k
# sort
actual.extend([
    np.nonzero(user_actual.numpy())[0] for user_actual in input_mask.cpu().detach() # input_mask -> valid_mask
])

# mask to train items
pred = pred * torch.logical_not(input_mask)
```
